app/src/main/java/com/sample/appconnectsample/ImageCaptureActivity.java

```diff
@@ -0,0 +1,193 @@
+            currentImage = (Bitmap) extras.get("data");
+            currentImageView.setImageBitmap(currentImage);
+            submitButton.setEnabled(true);
+        }
+        else if (requestCode == SELECT_IMAGE_INTENT && resultCode == RESULT_OK) {
+            Uri selectedImage = data.getData();
+            try {
+                InputStream imageStream = getContentResolver().openInputStream(selectedImage);
+                currentImage = BitmapFactory.decodeStream(imageStream);
+                currentImageView.setImageBitmap(currentImage);
+                submitButton.setEnabled(true);
+            } catch (FileNotFoundException exception) { }
+        }
+    }
+
+    private void dispatchSelectImageIntent() {
+        Intent photoPickerIntent = new Intent(Intent.ACTION_PICK);
+        photoPickerIntent.setType("image/*");
+        startActivityForResult(photoPickerIntent, SELECT_IMAGE_INTENT);
+    }
+
+    public void doSubmitButton(View source) {
+        submitButton.setEnabled(false);
+        new CollectAndSubmitTask().execute(currentImage);
+    }
+
+    /**
+     * An asynchronous task that collects the responses and submits the form.
+     */
+    private class CollectAndSubmitTask extends AsyncTask<Bitmap,Void,Void> {
+
+        private ProgressDialog progressDialog;
+        private String message;
+        private Exception exception;
+
+        @Override
+        protected void onPreExecute() {
+            progressDialog = new ProgressDialog(ImageCaptureActivity.this, ProgressDialog.STYLE_SPINNER);
+            progressDialog.setMessage(getString(R.string.image_capture_saving_message));
+            progressDialog.setCancelable(false);
+            progressDialog.show();
+        }
+
+        @Override
+        protected Void doInBackground(Bitmap... params) {
+            // *** AppConnect ***
+            // Each secondary thread must create its own datastore instance and
+            // dispose of it when done
+            Datastore datastore = null;
+
+            try {
+                datastore = DatastoreFactory.create();
+                Subject subject = datastore.getSubject(subjectID);
+
+                ByteArrayOutputStream stream = new ByteArrayOutputStream();
+                currentImage.compress(Bitmap.CompressFormat.JPEG, 70, stream);
+                byte[] byteArray = stream.toByteArray();
+
+                // the api call to babbage to actually save the data:
+                subject.collectData(byteArray, "sample metadata", "image/jpeg");
+            }
+            catch (Exception ex) {
+                exception = ex;
+                message = ex.getMessage();
+            }
+            finally {
+                if (datastore != null)
+                    datastore.dispose();
+            }
+            return null;
+        }
+
+        @Override
+        protected void onPostExecute(Void aVoid) {
+            progressDialog.cancel();
+
+            if (message != null) {
```

Comment:
Do the exception and the message both need to be stored? It might make more sense to just store the exception and then retrieve the message here.